src/contracts/test/SmartOrder.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.7.6;
+pragma abicoder v2;
+
+import "../interfaces/GPv2EIP1271.sol";
+import "../interfaces/IERC20.sol";
+import "../libraries/GPv2Order.sol";
+import "../libraries/GPv2SafeERC20.sol";
+import "../libraries/SafeMath.sol";
+import "../GPv2Settlement.sol";
+
+contract SmartOrder is EIP1271Verifier {
+    using GPv2Order for GPv2Order.Data;
+    using GPv2SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    bytes32 public constant APPDATA = keccak256("SmartOrder");
+
+    address public immutable owner;
+    bytes32 public immutable domainSeparator;
+    IERC20 public immutable sellToken;
+    IERC20 public immutable buyToken;
+    uint256 public immutable totalSellAmount;
+    uint256 public immutable totalFeeAmount;
+    uint32 public immutable validTo;
+
+    bool public funded;
+
+    constructor(
+        GPv2Settlement settlement,
+        IERC20 sellToken_,
+        IERC20 buyToken_,
+        uint32 validTo_,
+        uint256 totalSellAmount_,
+        uint256 totalFeeAmount_
+    ) {
+        owner = msg.sender;
+        domainSeparator = settlement.domainSeparator();
+        sellToken = sellToken_;
+        buyToken = buyToken_;
+        validTo = validTo_;
+        totalSellAmount = totalSellAmount_;
+        totalFeeAmount = totalFeeAmount_;
+
+        sellToken_.approve(
+            address(settlement.allowanceManager()),
+            totalSellAmount_.add(totalFeeAmount_)
+        );
+    }
+
+    modifier onlyOwner {
+        require(msg.sender == owner, "not owner");
+        _;
+    }
+
+    function fund() external onlyOwner {
+        require(!funded, "already funded");
+        funded = true;
+        sellToken.safeTransferFrom(
+            owner,
+            address(this),
+            totalSellAmount.add(totalFeeAmount)
+        );
+    }
+
+    function close() external onlyOwner {
+        uint256 balance = sellToken.balanceOf(address(this));
+        if (balance != 0) {
+            sellToken.safeTransfer(owner, balance);
+        }
+        selfdestruct(payable(owner));
+    }
+
+    function isValidSignature(bytes32 hash, bytes memory signature)
+        external
+        view
+        override
+        returns (bytes4 magicValue)
+    {
+        uint256 sellAmount = abi.decode(signature, (uint256));
+        GPv2Order.Data memory order = orderForSellAmount(sellAmount);
+
+        if (order.hash(domainSeparator) == hash) {
+            magicValue = GPv2EIP1271.MAGICVALUE;
+        }
+    }
+
+    function orderForSellAmount(uint256 sellAmount)
+        public
+        view
+        returns (GPv2Order.Data memory order)
+    {
+        uint256 balance = sellToken.balanceOf(address(this));
+        uint256 soldAmount =
+            totalSellAmount.sub(
+                balance.mul(totalSellAmount).div(
+                    totalSellAmount.add(totalFeeAmount)
+                )
+            );
```

Comment:
A small issue here is that if someone sends this contract some `sellToken` weis then no order can be executed anymore (funds can still be withdrawn). Maybe these lines can be moved to the `buyAmountForSellAmount` function and `balance` can be assumed tp be capped.

Reply:
Made this a saturating subtraction.